kvtenantccl: better address randomization in the connector

RaduBerinde opened the pull request:
#### kvtenantccl: better address randomization in the connector

The current dial code in the connector tries to dial all nodes in
circular order with randomized starting point.

This partial randomization is not quite what we want. For example,
if we have three addresses (0,1,2) and address 1 is down, two starting
points end up in choosing node 2 (so we choose it 67% of the time).

This commit changes to using `rand.Perm` instead. In the example
above, 0 and 2 have equal chances of being first in the permutation,
so we choose either of them 50% of the time.

Release justification: low risk change to existing functionality, only
affects multitenant (Serverless)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76815)
<!-- Reviewable:end -->

RaduBerinde commented:
Friendly ping

RaduBerinde commented:
TFTR!

bors r+

RaduBerinde commented:
bors r-

RaduBerinde committed:
kvtenantccl: better address randomization in the connector

The current dial code in the connector tries to dial all nodes in
circular order with randomized starting point.

This partial randomization is not quite what we want. For example,
if we have three addresses (0,1,2) and address 1 is down, two starting
points end up in choosing node 2 (so we choose it 67% of the time).

This commit changes to using `rand.Perm` instead. In the example
above, 0 and 2 have equal chances of being first in the permutation,
so we choose either of them 50% of the time.

Release justification: low risk change to existing functionality, only
affects multitenant (Serverless)

Release note: None

RaduBerinde commented:
bors r+

nvb reviewed (commented):
LGTM
